src/include/migraphx/op/pack_int4.hpp

```diff
@@ -62,35 +62,62 @@ struct pack_int4
     {
         check_shapes{inputs, *this}.same_dims().has(1);
         auto in_shape = inputs.front();
-        if(in_shape.type() != migraphx::shape::uint8_type)
+        if(in_shape.type() != migraphx::shape::int8_type and
+           in_shape.type() != migraphx::shape::uint8_type)
         {
-            MIGRAPHX_THROW("PACK_INT4: Only Unsigned Int8 type is supported for packing");
+            MIGRAPHX_THROW("PACK_INT4: Only Int8 or Uint8 is supported for packing");
         }
         auto new_lens = in_shape.lens();
         if(new_lens[axis] % 2 != 0)
         {
             MIGRAPHX_THROW("PACK_INT4: Can not pack axis that has odd lengths");
         }
         new_lens[axis] /= 2;
-        return {migraphx::shape::uint8_type, new_lens};
+        return {in_shape.type(), new_lens};
     }
 
     argument compute(const shape& output_shape, std::vector<argument> args) const
     {
+        auto input    = args.front();
+        auto in_shape = input.get_shape();
+
         argument result{output_shape};
-        auto in_shape = args.front().get_shape();
-        auto input    = args.at(0).get<uint8_t>();
-        auto output   = result.get<uint8_t>();
-        par_for(output_shape.elements(), [&](auto i) {
-            auto data_idx          = output_shape.multi(i);
-            auto in_data_multi_idx = data_idx;
-            in_data_multi_idx[axis] *= 2;
-            auto input_val = input[in_data_multi_idx];
-            // mask first 4 bits, keep it little endian.
-            output[i] = uint8_t(0x0F) & input_val;
-            in_data_multi_idx[axis] += 1;
-            input_val = input[in_data_multi_idx];
-            output[i] = (input_val << 4u) | output[i]; // NOLINT(hicpp-signed-bitwise)
+
+        visit_all(result, input)([&](auto out, auto inp) {
+            par_for(output_shape.elements(), [&](auto i) {
+                using type = typename decltype(inp)::value_type;
+                if constexpr(std::is_signed<type>{})
+                {
+                    const int8_t min_4bit  = -8; // clip min value
+                    const int8_t max_4bit  = 7;  // clip max value
+                    auto data_idx          = output_shape.multi(i);
+                    auto in_data_multi_idx = data_idx;
+                    in_data_multi_idx[axis] *= 2;
+                    int8_t val1 = inp[in_data_multi_idx];
+                    val1        = std::min(std::max(val1, min_4bit), max_4bit); // clip
+
+                    in_data_multi_idx[axis] += 1;
+                    int8_t val2 = inp[in_data_multi_idx];
+                    val2        = std::min(std::max(val2, min_4bit), max_4bit); // clip
+
+                    out[i] = (val2 << 4) | (val1 & 0xf); // pack
+                }
+                else
+                {
+                    const uint8_t max_4bit = 15; // clip max value
```

Comment:
only max and min values are differing. you can combine rest of the logic. You can set min value for unsigned as `0`

Reply:
You can also declare them using the `type` so it still uses the same type:

```cpp
type max_4bit = 0;
```

Reply:
Unfortunately, we get into issues with `visit_all`, where other types are getting involved in the compilation logic, and errors like this are showing up:
> error: use of overloaded operator '<<' is ambiguous (with operand types 'type' (aka 'half_float::half') and 'int')
And this is because it isn't just this line that is going to change.

Reply:
This code consolidation is taken care of. This constexpr is also avoided, and the code can now basically handle any type in visit_all clause here.